File: materials/orders-kotlin/app/src/main/kotlin/Processor.kt
// =========================================================
// Файл: Processor.kt
// Описание: Основная бизнес-логика.
// =========================================================

class OrderProcessor(
    private val database: RandomSQLDatabase,
    private val mailer: SmtpMailer
) {

    fun process(order: Order): Result<Unit> {
        println("--- Processing Order ${order.id} ---")

        // 1. Логика валидации
        if (order.items.isEmpty()) {
            return Result.failure(Exception("order must have at least one item"))
        }
        if (order.destination.city.isEmpty()) {
            return Result.failure(Exception("destination city is required"))
        }

        // 2. Логика расчета суммы
        var total = order.items.sumOf { it.price }

        // 3. Логика скидок и налогов
        when (order.type) {
            "Standard" -> {
                // Стандартный налог
                total *= 1.2
            }
            "Premium" -> {
                // Скидка 10% + налог
                total = (total * 0.9) * 1.2
            }
            "Budget" -> {
                if (order.items.size > 3) {
                    println("Budget orders cannot have more than 3 items. Skipping.")
                    return Result.success(Unit)
                }
            }
            "International" -> {
                total *= 1.5 // Таможенный сбор
                if (order.destination.city == "Nowhere") {
                    return Result.failure(Exception("cannot ship to Nowhere"))
                }
            }
            else -> return Result.failure(Exception("unknown order type"))
        }

        // 4. Логика сохранения
        runCatching {
            database.saveOrder(order, total)
        }.onFailure { e ->
            return Result.failure(Exception("database error: ${e.message}"))
        }

        // 5. Логика уведомлений
        val emailBody = "<h1>Your order ${order.id} is confirmed!</h1><p>Total: %.2f</p>".format(total)
        mailer.sendHtmlEmail(order.clientEmail, "Order Confirmation", emailBody)

        return Result.success(Unit)
    }
}

fun newOrderProcessor(): OrderProcessor {
    return OrderProcessor(
        database = newMySQLDatabase(),
        mailer = SmtpMailer(server = "smtp.google.com")
    )
}

File: materials/orders-python/processor.py
# processor.py
# =========================================================
# Файл: processor.py
# Описание: Основная бизнес-логика.
# =========================================================

from typing import Optional
from infrastructure import RandomSQLDatabase, SmtpMailer
from models import Order


class OrderProcessor:
    def __init__(self):
        self.database = RandomSQLDatabase()
        self.mailer = SmtpMailer()

    def process(self, order: Order) -> Optional[None]:
        print(f"--- Processing Order {order.id} ---")

        # 1. Валидация
        if not order.items or len(order.items) == 0:
            raise ValueError("order must have at least one item")
        if not order.destination.city:
            raise ValueError("destination city is required")

        # 2. Расчёт суммы
        total = 0.0
        for item in order.items:
            total += item.price

        # 3. Скидки и налоги
        if order.type == "Standard":
            total = total * 1.2
        elif order.type == "Premium":
            total = (total * 0.9) * 1.2
        elif order.type == "Budget":
            if len(order.items) > 3:
                print("Budget orders cannot have more than 3 items. Skipping.")
                return None
        elif order.type == "International":
            total = total * 1.5
            if order.destination.city == "Nowhere":
                raise ValueError("cannot ship to Nowhere")
        else:
            raise ValueError("unknown order type")

        # 4. Сохранение
        self.database.save_order(order, total)

        # 5. Уведомление
        email_body = f"<h1>Your order {order.id} is confirmed!</h1><p>Total: {total:.2f}</p>"
        self.mailer.send_html_email(order.client_email, "Order Confirmation", email_body)

        return None